pkg/kubelet/kubelet.go

```diff
@@ -2289,32 +2290,74 @@ func (kl *Kubelet) rejectPod(pod *api.Pod, reason, message string) {
+		// if kl.initialNodeStatus() returns with non-nil err, then the returned
+		// "node" is nil, which is not accepted for RunGeneralPredicates(). However,
+		// directly using `return kl.initialNodeStatus()` does not break any test for now.
+		// if n, err := kl.initialNodeStatus(); err == nil {
+		// 	return n, nil
+		// }
+		return kl.initialNodeStatus()
+	}
+	// if err != nil, n returns as nil which is not acceptable for RunGeneralPredicates()
+	if n, err := kl.nodeInfo.GetNodeInfo(kl.nodeName); err == nil {
+		return n, nil
+	}
+	// on failure, kl.kubeClient.Core().Nodes().Get(kl.nodeName) returns an empty &api.Node,
+	// which can be used by RunGeneralPredicates()
+	return kl.kubeClient.Core().Nodes().Get(kl.nodeName)
+}
+
 // canAdmitPod determines if a pod can be admitted, and gives a reason if it
 // cannot. "pod" is new pod, while "pods" include all admitted pods plus the
 // new pod. The function returns a boolean value indicating whether the pod
 // can be admitted, a brief single-word reason and a message explaining why
 // the pod cannot be admitted.
-//
-// This needs to be kept in sync with the scheduler's and daemonset's fit predicates,
-// otherwise there will inevitably be pod delete create loops. This will be fixed
-// once we can extract these predicates into a common library. (#12744)
 func (kl *Kubelet) canAdmitPod(pods []*api.Pod, pod *api.Pod) (bool, string, string) {
-	if hasHostPortConflicts(pods) {
-		return false, "HostPortConflict", "cannot start the pod due to host port conflict."
-	}
-	if !kl.matchesNodeSelector(pod) {
-		return false, "NodeSelectorMismatching", "cannot be started due to node selector mismatch"
-	}
-	cpu, memory := kl.hasInsufficientfFreeResources(pods)
-	if cpu {
-		return false, "InsufficientFreeCPU", "cannot start the pod due to insufficient free CPU."
-	} else if memory {
-		return false, "InsufficientFreeMemory", "cannot be started due to insufficient free memory"
+	node, err := kl.getNodeAnyWay()
+	if err != nil {
+		glog.Warningf("cannot get Node info, %v", err)
+		// no need to return "false" here, only RunGeneralPredicates() can result in denying a Pod
 	}
+	otherPods := []*api.Pod{}
+	for _, p := range pods {
+		if p != pod {
+			otherPods = append(otherPods, p)
+		}
+	}
+	nodeInfo := schedulercache.CreateNodeNameToInfoMap(otherPods)[kl.nodeName]
+	fit, err := predicates.RunGeneralPredicates(pod, kl.nodeName, nodeInfo, node)
+	if !fit {
+		if re, ok := err.(*predicates.PredicateFailureError); ok {
+			reason := re.PredicateName
+			message := re.Error()
+			glog.V(2).Infof("Predicate fails on Pod: %v, for reason: %v", format.Pod(pod), message)
+			return fit, reason, message
+		}
+		if re, ok := err.(*predicates.InsufficientResourceError); ok {
+			reason := re.ResourceName
+			message := re.Error()
+			glog.V(2).Infof("Predicate fails on Pod: %v, for reason: %v", format.Pod(pod), message)
+			return fit, reason, message
+		}
+		reason := "Unexpected predicate failure type"
```

Comment:
If this is being used for an event, then the reason has to declared in a central location with some documentation, and the reason has to contain only a single word.